### DISCUSSION TITLE
Using graphql api to create multi-line threaded comment at a specific commit

### DISCUSSION THREAD

**Discussion by ldelossa** [community/community#30317] [Apps, API and Webhooks]
Hello.

I'm writing [gh.nvim](https://github.com/ldelossa/gh.nvim) which is a Neovim plugin which integrates with the GitHub API. 

I'm trying to determine how to create a multiline threaded message on a particular commit. 

It's possible to make a non-multiline threaded comment with the `addPullRequestReviewComment` endpoint. This takes a `commitOID` as part of its input. This `commitOID` shows up in the comment's `originalCommit` field and all is well. 

However, that endpoint's field does not take a line range. The only graphql mutation endpoint which takes a line range is `addPullRequestReviewThread`. However, this endpoint does not take a `commitOID` and, to my knowledge, always takes its `body` field and creates an implicit comment which both it's `commit` and `originalCommit` fields pointing to the pull request's HEAD. 

This doesn't seem to be the case when I attempt this in the web UI. I can start an arbitrary review, go to a specific commit, and create a multiline object. What I receive back something similar to:

```
              node = {
                comments = {
                  edges = { {
                      node = {
                        author = {
                          avatarUrl = "https://avatars.githubusercontent.com/u/5642902?u=4ee288002460604dd7fc551750b0bfdef2c56ef5&v=4",
                          login = "ldelossa",
                          resourcePath = "/ldelossa",
                          url = "https://github.com/ldelossa"
                        },
                        body = "multiline comment",
                        commit = {
                          oid = "e6323cbce1fc884a88d47299d3bb070e4fcb6216",
                          parents = {
                            edges = { {
                                node = {
                                  id = "C_kwDOHIASLNoAKDg1Zjk3ZjhiOWUyYjNkNGQyMWM2MDM4MDQ4MmI2YzZkN2NjMDRjZjk",
                                  oid = "85f97f8b9e2b3d4d21c60380482b6c6d7cc04cf9"
                                }
                              } }
                          }
                        },
                        createdAt = "2022-08-21T20:26:12Z",
                        id = "PRRC_kwDOHIASLM44rYWC",
                        originalCommit = {
                          oid = "85f97f8b9e2b3d4d21c60380482b6c6d7cc04cf9",
                          parents = {
                            edges = { {
                                node = {
                                  id = "C_kwDOHIASLNoAKDg1Zjk3ZjhiOWUyYjNkNGQyMWM2MDM4MDQ4MmI2YzZkN2NjMDRjZjk",
                                  oid = "ae26b4fe7e1fe2c966d3a56cc61f73b0e4a1a07a"
                                }
                              } }
                          }
                        },
                        originalPosition = 5,
                        path = "README.md",
                        position = 5,
                        publishedAt = vim.NIL,
                        pullRequestReview = {
                          id = "PRR_kwDOHIASLM5AWx_l"
                        },
                        reactions = {
                          edges = {}
                        },
                        replyTo = vim.NIL,
                        state = "PENDING",
                        updatedAt = "2022-08-21T20:26:12Z",
                        url = "https://github.com/ldelossa/litee-gh-testing/pull/4#discussion_r950896002",
                        viewerCanDelete = true,
                        viewerCanMinimize = false,
                        viewerCanReact = true,
                        viewerCanUpdate = true,
                        viewerCannotUpdateReasons = {},
                        viewerDidAuthor = true
                      }
                    } }
                },
                diffSide = "RIGHT",
                id = "PRRT_kwDOHIASLM4kRFge",
                isOutdated = false,
                isResolved = false,
                line = 5,
                originalLine = 5,
                originalStartLine = 4,
                path = "README.md",
                resolvedBy = vim.NIL,
                startDiffSide = "RIGHT",
                startLine = 4,
                viewerCanReply = true,
                viewerCanResolve = false,
                viewerCanUnresolve = false
              }
            } },
          pageInfo = {
            endCursor = "Y3Vyc29yOnYyOpK0MjAyMi0wOC0yMVQyMDoyNjoxMlrOJERYHg==",
            hasNextPage = false
          }
```
As you can see, the thread has a comment created with an "originalCommit" pointing to the parent of that pull request's HEAD.

**Comment by iasemenov** [comment]
This is even more relevant with the deprecation/removal of diff-relative position APIs.
After the removal there will be no way to create a comment on a non-head commit.

**Reply by aleppard** [reply]
The AddPullRequestReviewComment has now been marked as deprecated:

```Upcoming Change on 2023-10-01 UTC Description: commitOID will be removed. use addPullRequestReviewThread or addPullRequestReviewThreadReply instead Reason: We are deprecating the addPullRequestReviewComment mutation.```

See https://docs.github.com/en/graphql/reference/input-objects#addpullrequestreviewthreadinput

The addPullRequestReviewThread does not take a commitOID argument. 

So it appears that there is currently no non-deprecated way of adding a comment to a specific commit - multi-line or not.

**Reply by aleppard** [reply]
I've raised a new discussion https://github.com/orgs/community/discussions/114345 as this affects all comments created through the GraphQL API now, and without a way of specifying the exact commit ID it opens up a race condition.

The workaround for non multi-threaded comments is now marked as deprecated.
